Add deployment type to aws_fsx_lustre_file_system

mtpdt opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Relates #12181

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
resource/aws_fsx_lustre_file_system: Add `deployment_type` and `per_unit_storage_throughput`
```

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make testacc TESTARGS='-run=TestAccAWSFsxLustreFileSystem_DeploymentType'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSFsxLustreFileSystem_DeploymentType -timeout 120m
make testacc TESTARGS='-run=TestAccAWSFsxLustreFileSystem_basic'
=== RUN   TestAccAWSFsxLustreFileSystem_DeploymentTypePersistent1
=== PAUSE TestAccAWSFsxLustreFileSystem_DeploymentTypePersistent1
=== RUN   TestAccAWSFsxLustreFileSystem_DeploymentTypeScratch2
=== PAUSE TestAccAWSFsxLustreFileSystem_DeploymentTypeScratch2
=== CONT  TestAccAWSFsxLustreFileSystem_DeploymentTypePersistent1
=== CONT  TestAccAWSFsxLustreFileSystem_DeploymentTypeScratch2
--- PASS: TestAccAWSFsxLustreFileSystem_DeploymentTypeScratch2 (434.35s)
--- PASS: TestAccAWSFsxLustreFileSystem_DeploymentTypePersistent1 (518.06s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	518.129s

$ make testacc TESTARGS='-run=TestAccAWSFsxLustreFileSystem_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSFsxLustreFileSystem_basic -timeout 120m
=== RUN   TestAccAWSFsxLustreFileSystem_basic
=== PAUSE TestAccAWSFsxLustreFileSystem_basic
=== CONT  TestAccAWSFsxLustreFileSystem_basic
--- PASS: TestAccAWSFsxLustreFileSystem_basic (615.14s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	615.195s
```

mtpdt commented:
Hello, this is my first pull request into `terraform-provider-aws`, so I appreciate any help I can get. In your [contribution guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/contributing/running-and-writing-acceptance-tests.md#acceptance-tests-often-cost-money-to-run), you mention an option to accept a "best effort" implementation of acceptance tests. Could you please try to run the acceptance tests I wrote for this feature?

DrFaust92 commented on aws/resource_aws_fsx_lustre_file_system.go:
```diff
@@ -108,6 +108,26 @@ func resourceAwsFsxLustreFileSystem() *schema.Resource {
 					validation.StringMatch(regexp.MustCompile(`^[1-7]:([01]\d|2[0-3]):?([0-5]\d)$`), "must be in the format d:HH:MM"),
 				),
 			},
+			"deployment_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					fsx.LustreDeploymentTypeScratch1,
+					fsx.LustreDeploymentTypeScratch2,
+					fsx.LustreDeploymentTypePersistent1,
+				}, false),
+			},
+			"per_unit_storage_throughput": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+				ValidateFunc: validation.IntInSlice([]int{
+					50,
+					100,
+					200,
+				}),
+			},
```
please add these fields to the read function.
something along the lines of:
```
d.Set("deployment_type", filesystem.LustreConfiguration.DeploymentType)
```

you would probably need to add default value of `fsx.LustreDeploymentTypeScratch1` to `deployment_type` as well.

mtpdt replied:
done, added a default and added the line you requested to the read function.

DrFaust92 commented on website/docs/r/fsx_lustre_file_system.html.markdown:
```diff
@@ -32,6 +32,8 @@ The following arguments are supported:
 * `security_group_ids` - (Optional) A list of IDs for the security groups that apply to the specified network interfaces created for file system access. These security groups will apply to all network interfaces.
 * `tags` - (Optional) A map of tags to assign to the file system.
 * `weekly_maintenance_start_time` - (Optional) The preferred start time (in `d:HH:MM` format) to perform weekly maintenance, in the UTC time zone.
+* `deployment_type` - (Optional) - The filesystem deployment type. For valid values, see the [AWS documentation](https://docs.aws.amazon.com/fsx/latest/APIReference/API_CreateFileSystemLustreConfiguration.html).
```
lets simplify this to enumerate the values in the docs here.

mtpdt replied:
Done, I followed the format I found from another AWS resource's doumentation.

DrFaust92 commented on aws/resource_aws_fsx_lustre_file_system_test.go:
```diff
@@ -382,6 +384,74 @@ func TestAccAWSFsxLustreFileSystem_WeeklyMaintenanceStartTime(t *testing.T) {
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"security_group_ids"},
+			},
+			{
```
no need for the second test as its a property that cannot be updated

mtpdt replied:
done

mtpdt commented:
@DrFaust92 requested changes implemented and unit tests pass.

I have a few more minor improvements to `aws_lustre_file_system` that will come after this PR. Would it help speed things along if I was able to run acceptance tests myself? I'm in a bit of a constrained environment, but if it's worth it, I could put work into being able to run the acceptance tests.

DrFaust92 commented on aws/resource_aws_fsx_lustre_file_system_test.go:
```diff
@@ -94,6 +94,8 @@ func TestAccAWSFsxLustreFileSystem_basic(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "0"),
 					resource.TestMatchResourceAttr(resourceName, "vpc_id", regexp.MustCompile(`^vpc-.+`)),
 					resource.TestMatchResourceAttr(resourceName, "weekly_maintenance_start_time", regexp.MustCompile(`^\d:\d\d:\d\d$`)),
+					resource.TestCheckResourceAttr(resourceName, "deployment_type", ""),
```
probably need to change it to the default value here

mtpdt replied:
done

DrFaust92 commented on aws/resource_aws_fsx_lustre_file_system_test.go:
```diff
@@ -382,6 +384,74 @@ func TestAccAWSFsxLustreFileSystem_WeeklyMaintenanceStartTime(t *testing.T) {
+		CheckDestroy: testAccCheckFsxLustreFileSystemDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAwsFsxLustreFileSystemDeploymentType("PERSISTENT_1"),
```
i think that it would be best to split this to 2 separate tests  for `PERSISTENT_1`
and just have a plan only test for `SCRATCH_1`
see examples here:
 https://github.com/terraform-providers/terraform-provider-aws/pull/12676/files#diff-17cef5d816ce52cd994aef2105a31a2dR117
https://github.com/terraform-providers/terraform-provider-aws/pull/12676/files#diff-17cef5d816ce52cd994aef2105a31a2dR124

mtpdt replied:
Done, I've created 2 separate tests, one for `PERSISTENT_1`, and another for `SCRATCH_2`.

And I've added a plan only test for `SCRATCH_1`.

DrFaust92 commented:
@mtpdt, some more changes to tests and it would be best if you could run these to save us sometime and cycles

mtpdt commented:
@DrFaust92 thank you for the reviews. All current comments addressed.

I'll work on my end to run the tests myself for the subsequent PRs I send.

DrFaust92 commented:
getting failures for a few tests:
```
TestAccAWSFsxLustreFileSystem_PerUnitStorageThroughput
2020/07/22 13:29:52 [ERROR] <root>: eval: *terraform.EvalApplyPost, err: Error creating FSx filesystem: IncompatibleParameterError: The storage throughput can not be specified for a lustre file system with DeploymentType=SCRATCH_2

    TestAccAWSFsxLustreFileSystem_basic: testing.go:684: Step 0 error: Check failed: Check 16/16 error: aws_fsx_lustre_file_system.test: Attribute 'per_unit_storage_throughput' not found

TestAccAWSFsxLustreFileSystem_Scratch2DeploymentType: testing.go:684: Step 0 error: errors during apply:
        
        Error: Error creating FSx filesystem: IncompatibleParameterError: The storage throughput can not be specified for a lustre file system with DeploymentType=SCRATCH_2
```

it seems that `per_unit_storage_throughput` can only be set with a persistent deployment type

DrFaust92 commented:
actually it seems you are missing something along the lines of the following in the read func:

```
	if filesystem.LustreConfiguration.PerUnitStorageThroughput != nil {
		d.Set("per_unit_storage_throughput", filesystem.LustreConfiguration.PerUnitStorageThroughput)
	}
```

mtpdt commented:
@DrFaust92 thank you for that code snippet.

I've fixed both errors, and I've managed to get the acceptance tests running in my environment. I've updated the PR comment to include the passing acceptance test output for the new tests I've added, as well as the basic integration test for fsx lustre.

mtpdt commented on aws/resource_aws_fsx_lustre_file_system_test.go:
```diff
@@ -382,6 +387,48 @@ func TestAccAWSFsxLustreFileSystem_WeeklyMaintenanceStartTime(t *testing.T) {
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckFsxLustreFileSystemExists(resourceName, &filesystem),
+					// per_unit_storage_throughput is only available with deployment_type=PERSISTENT_1, so we test both here.
+					resource.TestCheckResourceAttr(resourceName, "per_unit_storage_throughput", "50"),
```
I took out the separate per unit storage throughput test, and combined it into this one, since per unit storage throughput is only valid for persistent1 type lustre.

DrFaust92 commented on aws/resource_aws_fsx_lustre_file_system_test.go:
```diff
@@ -382,6 +387,48 @@ func TestAccAWSFsxLustreFileSystem_WeeklyMaintenanceStartTime(t *testing.T) {
+					resource.TestCheckResourceAttr(resourceName, "per_unit_storage_throughput", "50"),
+					resource.TestCheckResourceAttr(resourceName, "deployment_type", fsx.LustreDeploymentTypePersistent1),
+				),
+			},
```
can you add:
```
{
				ImportStateVerify:       true,
				ImportStateVerifyIgnore: []string{"security_group_ids"},
			},
```

to the added tests?

DrFaust92 commented:
thanks @mtpdt, i'll try to run these later today. fsx take so long to create and test 😢

DrFaust92 commented:
sorry @mtpdt, I messed up the import test addition due to copy paste

can you change it to 

```
			{
				ResourceName:            resourceName,
				ImportState:             true,
				ImportStateVerify:       true,
				ImportStateVerifyIgnore: []string{"security_group_ids"},
			},
```


otherwise looks good and well run tests one more time and we are good to go

mtpdt commented:
@DrFaust92 no problem, thank you for running the tests for me. 

Note that after this one, I have two much smaller related PRs: [14314](https://github.com/terraform-providers/terraform-provider-aws/pull/14314) and [14313](https://github.com/terraform-providers/terraform-provider-aws/pull/14313), I have already run the acceptance tests for both of them.

breathingdust commented:
LGTM 🚀 Thanks @DrFaust92!

Verified Acceptance Tests

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSFsxLustreFileSystem_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSFsxLustreFileSystem_ -timeout 120m
=== RUN   TestAccAWSFsxLustreFileSystem_basic
=== PAUSE TestAccAWSFsxLustreFileSystem_basic
=== RUN   TestAccAWSFsxLustreFileSystem_disappears
=== PAUSE TestAccAWSFsxLustreFileSystem_disappears
=== RUN   TestAccAWSFsxLustreFileSystem_ExportPath
=== PAUSE TestAccAWSFsxLustreFileSystem_ExportPath
=== RUN   TestAccAWSFsxLustreFileSystem_ImportPath
=== PAUSE TestAccAWSFsxLustreFileSystem_ImportPath
=== RUN   TestAccAWSFsxLustreFileSystem_ImportedFileChunkSize
=== PAUSE TestAccAWSFsxLustreFileSystem_ImportedFileChunkSize
=== RUN   TestAccAWSFsxLustreFileSystem_SecurityGroupIds
=== PAUSE TestAccAWSFsxLustreFileSystem_SecurityGroupIds
=== RUN   TestAccAWSFsxLustreFileSystem_StorageCapacity
=== PAUSE TestAccAWSFsxLustreFileSystem_StorageCapacity
=== RUN   TestAccAWSFsxLustreFileSystem_Tags
=== PAUSE TestAccAWSFsxLustreFileSystem_Tags
=== RUN   TestAccAWSFsxLustreFileSystem_WeeklyMaintenanceStartTime
=== PAUSE TestAccAWSFsxLustreFileSystem_WeeklyMaintenanceStartTime
=== RUN   TestAccAWSFsxLustreFileSystem_DeploymentTypePersistent1
=== PAUSE TestAccAWSFsxLustreFileSystem_DeploymentTypePersistent1
=== RUN   TestAccAWSFsxLustreFileSystem_DeploymentTypeScratch2
=== PAUSE TestAccAWSFsxLustreFileSystem_DeploymentTypeScratch2
=== CONT  TestAccAWSFsxLustreFileSystem_basic
=== CONT  TestAccAWSFsxLustreFileSystem_StorageCapacity
=== CONT  TestAccAWSFsxLustreFileSystem_DeploymentTypePersistent1
=== CONT  TestAccAWSFsxLustreFileSystem_WeeklyMaintenanceStartTime
=== CONT  TestAccAWSFsxLustreFileSystem_Tags
=== CONT  TestAccAWSFsxLustreFileSystem_ExportPath
=== CONT  TestAccAWSFsxLustreFileSystem_DeploymentTypeScratch2
=== CONT  TestAccAWSFsxLustreFileSystem_ImportPath
=== CONT  TestAccAWSFsxLustreFileSystem_SecurityGroupIds
=== CONT  TestAccAWSFsxLustreFileSystem_ImportedFileChunkSize
=== CONT  TestAccAWSFsxLustreFileSystem_disappears
--- PASS: TestAccAWSFsxLustreFileSystem_DeploymentTypePersistent1 (426.79s)
--- PASS: TestAccAWSFsxLustreFileSystem_WeeklyMaintenanceStartTime (440.55s)
--- PASS: TestAccAWSFsxLustreFileSystem_basic (486.00s)
--- PASS: TestAccAWSFsxLustreFileSystem_Tags (541.28s)
--- PASS: TestAccAWSFsxLustreFileSystem_DeploymentTypeScratch2 (548.79s)
--- PASS: TestAccAWSFsxLustreFileSystem_disappears (559.20s)
--- PASS: TestAccAWSFsxLustreFileSystem_StorageCapacity (996.81s)
--- PASS: TestAccAWSFsxLustreFileSystem_SecurityGroupIds (1010.93s)
--- PASS: TestAccAWSFsxLustreFileSystem_ImportedFileChunkSize (1214.90s)
--- PASS: TestAccAWSFsxLustreFileSystem_ExportPath (1457.39s)
--- PASS: TestAccAWSFsxLustreFileSystem_ImportPath (1476.03s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1478.540s
```

mtpdt commented:
nice, thanks!
